### ISSUE TITLE
Support deleting RRef in circular dependencies

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#27096]
This was original mentioned by @aazzolini ([comment](https://github.com/pytorch/pytorch/pull/25499#discussion_r328411474))

The implementation in #25499 does not delete RRef correctly when it is inside some object with circular dependency. The following code would expose the leak.

```python
    @_wrap_with_rpc
    def test_circular_dependency(self):
        n = self.rank + 1
        dst_rank = n % self.world_size
        class EmptyClass:
            pass
        a = EmptyClass()
        b = EmptyClass()
        a.other = b
        b.other = a
        a.rref = dist.remote(
            'worker{}'.format(dst_rank),
            torch.add,
            args=(torch.ones(n, n), 2)
        )

```

Ref: 

- Python GC. https://rushter.com/blog/python-garbage-collector/
- Python Ref Count C API. https://pythonextensionpatterns.readthedocs.io/

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528

**Comment by mrshenli** [comment]
Discussed with @xush6528 offline and we are trying to design a solution for this. We could extend  #31893 to solve both this issue and #31325.

### Motivation

Currently, we already have a `_ignore_leaks` mode for non-graceful shutdown, and we would like to reliably delete RRefs for the graceful shutdown mode.

### Problem Statement

Following things could lead to RRef leaks:

1. One object in a local circular dependency that points to a `UserRRef`, as described in this issue.  
2. One object in a distributed circular dependency that points to a `UserRRef`. E.g., a `UserRRef` `R_A` on node A points to a list `L_B` on node B, where `L_B` contains a `UserRRef` `R_B` that points to a lis `L_A` on A, and `L_A` contains `R_A`
3. Python GC didn't run in time before `shutdown` is called.
4. Non-collective `RpcAgent` that cannot use sync and barrier to flush all pending system messages.

### Solution

1. Use `_wait_all_workers()` to flush all application RPC messages. @xush6528 has implemented this in #32190
2. Use `_destroy_rref_context()` to block until all RRef-related system messages are processed.

More specifically, each `UserRRef` has the following states:

* From Owner-User perspective: unconfirmed -> confirmed -> deleted
* From Parent-Child perspective: forked -> accepted ->delete

The `RRefContext` can then 

1. track all UserRRefs using weak pointers as already implemented in #31893.
2. When calling `shutdown`, for confirmed and accepted `UserRRef`s, we can send the delete message and delete locally right away. 
3. If a `UserRRef` is either not confirmed or not accepted, skip it.
4. If there are still pending `UserRRef`s or pending `OwnerRRef`s in a `RRefContext`, `_destroy_rref_context()` would block waiting. When future `REMOTE_RET`, `RREF_USER_DELETE`, or `RREF_CHILD_ACCEPT` arrives, they can wake up the `RRefContext` to check again.

Note that, as `_wait_all_workers()` has flushed user RPC messages, there won't be any additional RRef forks.

**Comment by mrshenli** [comment]
closed by #31893
